core/ecschema-editing/src/Editing/Editor.ts

```diff
@@ -111,69 +86,65 @@ export class SchemaContextEditor {
     await this._schemaContext.addSchema(newSchema);
-    return { schemaKey: newSchema.schemaKey };
+    return newSchema.schemaKey;
   }
 
   /**
    * Adds a referenced schema to the schema identified by the given SchemaKey.
    * @param schemaKey The SchemaKey identifying the schema.
    * @param refSchema The referenced schema to add.
    */
-  public async addSchemaReference(schemaKey: SchemaKey, refSchema: Schema): Promise<SchemaEditResults> {
+  public async addSchemaReference(schemaKey: SchemaKey, refSchema: Schema): Promise<void> {
     const schema = (await this.schemaContext.getCachedSchema<MutableSchema>(schemaKey, SchemaMatchType.Exact));
     if (schema === undefined)
-      return { errorMessage: `Schema Key ${schemaKey.toString(true)} not found in context` };
+      throw new ECEditingError(ECEditingStatus.SchemaNotFound, `Schema Key ${schemaKey.toString(true)} not found in context`);
 
     await schema.addReference(refSchema);
-    const diagnostics = Rules.validateSchemaReferences(schema);
+    const diagnosticIterable = Rules.validateSchemaReferences(schema);
 
-    const result: SchemaEditResults = { errorMessage: "" };
-    for await (const diagnostic of diagnostics) {
-      result.errorMessage += `${diagnostic.code}: ${diagnostic.messageText}\r\n`;
+    const diagnostics: AnyDiagnostic[] = [];
+    for await (const diagnostic of diagnosticIterable) {
+      diagnostics.push(diagnostic);
     }
 
-    if (result.errorMessage) {
+    if (diagnostics.length > 0) {
       this.removeReference(schema, refSchema);
-      return result;
+      throw new ECEditingError(ECEditingStatus.RuleViolation, undefined, diagnostics);
     }
 
     if (!await this.schemaContext.getCachedSchema(refSchema.schemaKey)) {
       await this.schemaContext.addSchema(refSchema);
     }
-
-    return {};
   }
 
   /**
    * Adds a CustomAttribute instance to the schema identified by the given SchemaKey
    * @param schemaKey The SchemaKey identifying the schema.
    * @param customAttribute The CustomAttribute instance to add.
    */
-  public async addCustomAttribute(schemaKey: SchemaKey, customAttribute: CustomAttribute): Promise<SchemaEditResults> {
+  public async addCustomAttribute(schemaKey: SchemaKey, customAttribute: CustomAttribute): Promise<void> {
     const schema = (await this.schemaContext.getCachedSchema<MutableSchema>(schemaKey, SchemaMatchType.Latest));
     if (schema === undefined)
-      return { errorMessage: `Schema Key ${schemaKey.toString(true)} not found in context` };
+      throw new ECEditingError(ECEditingStatus.SchemaNotFound, `Schema Key ${schemaKey.toString(true)} not found in context`);
 
     schema.addCustomAttribute(customAttribute);
 
-    const diagnostics = Rules.validateCustomAttributeInstance(schema, customAttribute);
+    const diagnosticIterable = Rules.validateCustomAttributeInstance(schema, customAttribute);
```

Comment:
Since the iterable is only been used in the for loop in line 140, we should probably just use it there:
`for await (const diagnostic of Rules.validateCustomAttributeInstance(schema, customAttribute)) {`